Add datamodule parameter to lr_find()

GimmickNG opened the pull request:
<!--
Please note that we have freeze state on adding new features till v1.0 release.
Anyway, any new feature suggestion is welcome and you are free to draft a PR,
 but be aware that several refactoring will take place for this major release yielding in significant collision solving.
A recommendation for feature draft is draw just outline, do not implement all details and propagate the changes to codebase...
-->

## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
 Please also include relevant motivation and context.
 List any dependencies that are required for this change.
-->

Fixes #3424 - Adds datamodule support to ``Trainer.lr_find()`` - a ``LightningDataModule`` can be passed to ``lr_find()``.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

williamFalcon commented:
does this work?
can you add a test?

GimmickNG commented:
Yeah, it does - AFAIK all it does is forward the ``datamodule`` parameter to the ``fit()`` method for the temporary trainer created in ``lr_find()``. 

I'm not quite sure how to write tests but I'll write them once I figure out how to do so.

SkafteNicki commented:
@GimmickNG changes look fine.
For testing, you can write a altered version of this test:
https://github.com/PyTorchLightning/pytorch-lightning/blob/5abf7d91232322e7f87fe65ad490245d9c1894c4/tests/trainer/test_lr_finder.py#L133-L152
where you remove all dataloaders in the model and then feed in a datamodule to the `lr_find` method. If the search finish the data is passed correctly.

Borda commented on pytorch_lightning/trainer/lr_finder.py:
```diff
@@ -98,6 +98,7 @@ def lr_find(
             model: LightningModule,
             train_dataloader: Optional[DataLoader] = None,
             val_dataloaders: Optional[Union[DataLoader, List[DataLoader]]] = None,
+            datamodule: Optional[LightningDataModule] = None,
```
lets move it at the end so we do not break compatibility...

GimmickNG replied:
Done, thanks for pointing it out.

GimmickNG commented:
> 
> 
> @GimmickNG changes look fine.
> For testing, you can write a altered version of this test:
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/5abf7d91232322e7f87fe65ad490245d9c1894c4/tests/trainer/test_lr_finder.py#L133-L152
> 
> where you remove all dataloaders in the model and then feed in a datamodule to the `lr_find` method. If the search finish the data is passed correctly.

I've added a test for testing the ``datamodule`` parameter. I'm not sure if I've done it correctly, though.

[I'd written a test which checked whether passing the ``datamodule`` parameter works when calling ``trainer.tune(model)`` with ``Trainer(auto_lr_find=True)``, because I was under the impression that calling ``tune()`` would be different to calling ``tuner.lr_find()`` - I've rolled the changes back and simplified the test to the one suggested by @SkafteNicki]

SkafteNicki commented on tests/trainer/test_lr_finder.py:
```diff
@@ -152,6 +152,31 @@ def test_call_to_trainer_method(tmpdir):
+        max_epochs=2,
+    )
+
+    lrfinder = trainer.lr_find(model, mode='linear')
```
```suggestion
    lrfinder = trainer.tuner.lr_find(model, datamodule=dm)
```

GimmickNG replied:
Oops, I remembered a short while after committing that I'd forgotten to add the ``datamodule`` parameter. That'll teach me to copy-paste code...

SkafteNicki commented on tests/trainer/test_lr_finder.py:
```diff
@@ -152,6 +152,31 @@ def test_call_to_trainer_method(tmpdir):
+    lrfinder = trainer.lr_find(model, mode='linear')
+    after_lr = lrfinder.suggestion()
+    model.learning_rate = after_lr
+    trainer.tune(model)
```
```suggestion

```
This is not necessary in the test, since we call the method directly

GimmickNG replied:
I see, alright

SkafteNicki commented:
We are currently doing a lot of internal refactors and the file `pytorch_lightning/trainer/lr_finder.py` therefore do not exist anymore (the conflicted file). Please redo the changes to this file: `pytorch_lightning/tuner/lr_finder.py`.

edenlightning commented:
Hey @GimmickNG! Refactors are mostly done, mind taking another look?

GimmickNG commented:
> 
> 
> Hey @GimmickNG! Refactors are mostly done, mind taking another look?

Sorry, I've been rather busy the past week. Should I create another pull request with the updated version of the fork?

Borda commented:
> > Hey @GimmickNG! Refactors are mostly done, mind taking another look?
> 
> Sorry, I've been rather busy the past week. Should I create another pull request with the updated version of the fork?

I think that resolving conflict is fine (rebase on actual master)

GimmickNG commented:
Sorry, I can't seem to do that on the web client and I'm rather inexperienced with this. How would I get Git to recognize the file's been moved to a different module during the rebase?

SkafteNicki commented:
@GimmickNG I merged master into your code and fixed conflicts. Failing checks seems to be unrelated to PR. Please take a look and make sure the PR is still doing what you intended, then we can go ahead and approve changes.

carmocca commented on pytorch_lightning/tuner/lr_finder.py:
```diff
@@ -98,6 +101,9 @@ def lr_find(
             loss at any point is larger than early_stop_threshold*best_loss
             then the search is stopped. To disable, set to None.
 
+        datamodule: A instance of Pytorch lightning data module
```
```suggestion
        datamodule: A :class:`LightningDataModule`.
```

GimmickNG commented:
> 
> 
> @GimmickNG I merged master into your code and fixed conflicts. Failing checks seems to be unrelated to PR. Please take a look and make sure the PR is still doing what you intended, then we can go ahead and approve changes.

Thanks! I checked it and ``trainer.tuner.lr_find(model, datamodule=data_module)`` works, at least on my end. I think it should be fine to close.

Borda commented on pytorch_lightning/tuner/tuning.py:
```diff
@@ -60,7 +62,8 @@ def lr_find(
             mode,
-            early_stop_threshold
+            early_stop_threshold,
+            datamodule
```
```suggestion
            datamodule,
```

Borda reviewed (approved):
LGTM

Borda commented:
> > @GimmickNG I merged master into your code and fixed conflicts. Failing checks seems to be unrelated to PR. Please take a look and make sure the PR is still doing what you intended, then we can go ahead and approve changes.
> 
> Thanks! I checked it and `trainer.tuner.lr_find(model, datamodule=data_module)` works, at least on my end. I think it should be fine to close.

I would keep it for merge as it adds test for the before/after...

justusschock commented on pytorch_lightning/tuner/lr_finder.py:
```diff
@@ -98,6 +101,9 @@ def lr_find(
             loss at any point is larger than early_stop_threshold*best_loss
             then the search is stopped. To disable, set to None.
 
+        datamodule: A :class:`LightningDataModule`.
```
this docstring is not saying anything meaningful. Instead of the class (which is derived from type annotation in docs) could you maybe add a short description here what this parameter does?

GimmickNG replied:
Sorry for the delay. Added a brief description, although it's rather self-explanatory because I couldn't find any matching documentation on ``trainer.fit()`` to reference.

SkafteNicki reviewed (approved):
LGTM

pep8speaks commented:
Hello @GimmickNG! Thanks for updating this PR. 

* In the file [`pytorch_lightning/tuner/lr_finder.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/aa5c913ddd7deedfdcce37cb5d8461c6e957692e/pytorch_lightning/tuner/lr_finder.py):

> [Line 105:79](https://github.com/PyTorchLightning/pytorch-lightning/blob/aa5c913ddd7deedfdcce37cb5d8461c6e957692e/pytorch_lightning/tuner/lr_finder.py#L105): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

williamFalcon commented:
Please rebae master...
cc @Borda
